RabbitMQ formula for installing and managing server and clustering

blarghmatey opened the pull request:
Closes mitodl/configuration#74

blarghmatey committed:
Installing RabbitMQ working and tested

Added a plan for installing RabbitMQ with its dependencies. Used
testinframod for writing a test plan to validate installation.

Addresses mitodl/salt-ops#74

blarghmatey committed:
Added support for installing plugins

Added a state file for installing and enabling plugins, along with
associated test states.

blarghmatey committed:
Added `configure` state file

Added a state plan for generating an Erlang style configuration file for
RabbitMQ and set the recommended memory and disk limits according to the
resources available on the machine.

blarghmatey committed:
Added additional states for config and tests

Added state files for handling:
- generation of RabbitMQ configuration file
- setting Erlang cookie
- adding users, vhosts and policies based on pillar data
- setting file descriptor limit
- testing proper configuration of above

blarghmatey committed:
Added generation of rabbitmq-env file

Added state to generate rabbitmq-env file in order to use the
[rabbitmq-clusterer](https://github.com/rabbitmq/rabbitmq-clusterer) plugin.

Refactored `configure.sls` to use the `pyobjects` DSL instead of the
`pydsl` because of the fact that `pyobjects` is more up to date,
maintained, and idiomatic.

Updated requirements ordering to ensure proper execution out of the box.

blarghmatey committed:
Updated README with new state files

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+#!pyobjects
```
Was there a particular reason this needed to be a pyobjects instead of the default jinja-yaml pipeline? my thoughts are that it better to write `determine_(ram|disk)_limit`.

blarghmatey replied:
The reason for it being done as pyobjects is because of the calculations necessary for determining the ram and disk limit values. I'm not sure how I would go about importing the results of the computation from a pyobjects file to a jinja/yaml file.

bdero replied:
Yeah, that's what I figured, thanks for confirming. :+1:

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+            d[k] += v
+        else:
+            d[k] = u[k]
+    return d
```
Can you add a comment to link back to this for documentation purposes? http://stackoverflow.com/a/3233356/5221054

It took me some time to realize why it was necessary.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+            socket.inet_pton(socket.AF_INET6, addr_string)
+        except (AttributeError, OSError, socket.error):
+            return False
+    return True
```
Cool way to validate an address using python's batteries - it's simple but I haven't seen it before.

blarghmatey replied:
Yeah, I had to google around for that one since the ipaddress module isn't present in 2.7.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+    limit = {'mem_relative': max(set_ratio, min_ratio)}
+    if space_override:
+        limit = total_disk * max_ratio
+    return limit
```
Did you use any external resources to come up with these two calculations? It seems worth pointing to any doc(s) that reason about this.

blarghmatey replied:
These are based on the documentation for RabbitMQ found [here](http://www.rabbitmq.com/production-checklist.html). Adding that in as a comment now.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+        config += ']},\n'
+    config = config.strip(',\n')
+    config += '].'
+    return config
```
Just a thought: Seems like it would be cool to make something like this a util in salt along with an erlang jinja2 filter.

blarghmatey replied:
Agreed.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+
+rabbit_pillar_conf = pillar('rabbitmq:configuration', {})
+
+rabbit_conf = update(rabbit_conf, rabbit_pillar_conf)
```
Was going to say that maybe the erlang config generator isn't worth it, but I feel like this makes it worth it.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+                 group='rabbitmq',
+                 contents=salt.hashutil.md5_digest(
+                     str(pillar('rabbitmq_configuration',
+                                {})).lower()),
```
Ok, taking a hash of some per-stack thing seems good, but it looks like this isn't in the pillar example (unless maybe `_` maps to `.` or something else I don't know). Should it be?

blarghmatey replied:
That's a typo, good catch. `_` should be `:`. Fixing it now.

bdero commented on rabbitmq/configure.sls:
```diff
@@ -0,0 +1,128 @@
+    Cmd.wait('stop_erlang_vm',
+             name='pkill beam',
+             watch=[File('set_rabbitmq_erlang_cookie'),
+                    File('generate_rabbitmq_env_file')])
```
Is this the correct way to restart erlang, or is it maybe a workaround for something?

blarghmatey replied:
The rabbitmq-server command starts up the erlang VM but then when trying to call `systemctl restart rabbitmq-server` it uses `rabbitmqctl` which doesn't affect the state of the underlying erlang VM. In order to be able to change the erlang cookie and the erlang env variables the VM itself needs to be restarted. Since there is no service management layer for the VM specifically it is necessary to use the kill command.

bdero replied:
Ah, ok, thanks for walking me though that.

blarghmatey committed:
Added informative docstrings, typo fix

Added docstrings and comments to provide more context around operations
in configure state. Fixed typo in reference for pillar data lookup.

bdero commented on rabbitmq/os_tweaks.sls:
```diff
@@ -0,0 +1,11 @@
+    - name: sysctl -w fs.file-max={{ rabbitmq.fd_limit }}
+  file.append:
+    - name: /etc/sysctl.conf
+    - text: fs.file_max={{ rabbitmq.fd_limit }}
```
Is this for setting the max to be high due to rabbitmq potentially open streams for a large number of files?

blarghmatey replied:
Yes, this is because of the need to potentially have a large number of open network sockets and for when messages need to get paged to disk. http://www.rabbitmq.com/production-checklist.html

bdero commented on rabbitmq/plugins.sls:
```diff
@@ -0,0 +1,31 @@
+download_rabbitmq_plugin_{{ plugin.name }}:
+  file.managed:
+    - name: /usr/lib/rabbitmq/lib/rabbitmq_server-{{ rabbitmq.version.split('-')[0] }}/plugins/{{ plugin.package_name }}
+    - source: https://www.rabbitmq.com/community-plugins/v3.6.x/{{ plugin.package_name }}
```
Hmm do you think it'd be safe to pull out the `major.minor` section from `rabbitmq.version` here and plop it in this URL like `/v{{ ...cutout... }}.x/`? My thought is it would be a more explicit failure on incompatible plugins if it fails to download plugins on these steps after an upgrade.

blarghmatey replied:
Yeah, probably worth doing that. I can add a fix unless you want to try doing a PR to patch it after we merge.

bdero commented on scripts/vagrant_setup.sh:
```diff
@@ -10,12 +10,13 @@ sudo mkdir -p /srv/salt
 sudo mkdir -p /srv/formulas
 sudo cp $SRCDIR/pillar.example /srv/pillar/pillar.sls
-sudo cp -r $SRCDIR/rabbitmq /srv/salt
+sudo ln -s $SRCDIR/rabbitmq /srv/salt/rabbitmq
```
Hah, maybe we should be doing this for our other formula dev environments. I saw that you also turned on rsync.

blarghmatey replied:
Yes, I agree. It is a more useful and robust way for managing that mapping.

bdero commented on scripts/validate_erlang_config.erl:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env escript
+main([ConfigFile]) ->
+    {ok, Terms} = file:consult(ConfigFile),
+    io:format("~p~n",[Terms]).
```
Is this for manual testing purposes, or does something execute this that I'm not seeing?

blarghmatey replied:
Yes, this was just for manually validating that the config file generated by the function I wrote was correct and able to be interpreted by Erlang.

bdero commented:
I just had a couple of more questions and one suggestion/discussion. After that, :+1: .
